src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageActionDescriptorProvider.cs

```diff
@@ -0,0 +1,125 @@
+                {
+                    throw new InvalidOperationException(string.Format(
+                        Resources.PageActionDescriptorProvider_RouteTemplateCannotBeOverrideable,
+                        item.Path));
+                }
+
+                AddActionDescriptors(context.Results, item, template);
+            }
+        }
+
+        public void OnProvidersExecuted(ActionDescriptorProviderContext context)
+        {
+        }
+
+        private void AddActionDescriptors(IList<ActionDescriptor> actions, RazorProjectItem item, string template)
+        {
+            var model = new PageModel(item.CombinedPathWithoutExtension, item.PathWithoutExtension);
+            model.Selectors.Add(CreateSelectorModel(item.CombinedPathWithoutExtension, template));
+
+            if (string.Equals(IndexFileName, item.Filename, StringComparison.OrdinalIgnoreCase))
+            {
+                model.Selectors.Add(CreateSelectorModel(item.BasePath, template));
+            }
+
+            for (var i = 0; i < _pagesOptions.Conventions.Count; i++)
+            {
+                _pagesOptions.Conventions[i].Apply(model);
+            }
+
+            var filters = new List<FilterDescriptor>(_mvcOptions.Filters.Count + model.Filters.Count);
+            for (var i = 0; i < _mvcOptions.Filters.Count; i++)
+            {
+                filters.Add(new FilterDescriptor(_mvcOptions.Filters[i], FilterScope.Global));
+            }
+
+            for (var i = 0; i < model.Filters.Count; i++)
+            {
+                filters.Add(new FilterDescriptor(model.Filters[i], FilterScope.Action));
+            }
+
+            foreach (var selector in model.Selectors)
+            {
+                actions.Add(new PageActionDescriptor()
+                {
+                    AttributeRouteInfo = new AttributeRouteInfo()
+                    {
+                        Name = selector.AttributeRouteModel.Name,
+                        Order = selector.AttributeRouteModel.Order ?? 0,
+                        Template = selector.AttributeRouteModel.Template,
+                    },
+                    DisplayName = $"Page: {item.Path}",
+                    FilterDescriptors = filters,
+                    Properties = new Dictionary<object, object>(model.Properties),
+                    RelativePath = item.CombinedPath,
+                    RouteValues = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
+                    {
+                        { "page", item.PathWithoutExtension },
+                    },
+                    ViewEnginePath = item.Path,
+                });
+            }
+        }
+
+        private static SelectorModel CreateSelectorModel(string prefix, string template)
+        {
+            return new SelectorModel
+            {
+                AttributeRouteModel = new AttributeRouteModel
+                {
+                    Template = AttributeRouteModel.CombineTemplates(prefix, template),
```

Comment:
I think you can drastically simplify this. As far as I understand, prefix is never going to be null, and template does not support override patterns so you can just do `template != null ? (prefix.EndsWith("/") ? prefix + template : $"{prefix}/{template}") : prefix;`

I think we shouldn't be using `AttributeRouteModel.CombineTemplates` here. It is basically for describing how to combine routes when created through attributes on the controller and action. Reusing it here feels wrong.